File: govuknotify1/src/main/java/uk/gov/nationalarchives/tdr/govuknotify/NotifyEmailSenderProvider.java
package uk.gov.nationalarchives.tdr.govuknotify;

import org.keycloak.email.EmailSenderProvider;
import org.keycloak.models.UserModel;
import uk.gov.service.notify.NotificationClient;
import uk.gov.service.notify.NotificationClientException;

import java.util.AbstractMap;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class NotifyEmailSenderProvider implements EmailSenderProvider {

    private NotificationClient notifyClient;

    public NotifyEmailSenderProvider(NotificationClient notifyClient) {
        this.notifyClient = notifyClient;
    }

    @Override
    public void send(Map<String, String> config, UserModel user, String subject, String textBody,
                     String htmlBody) {

        Map<String, ?> personalisation = Stream.of(
                new AbstractMap.SimpleImmutableEntry<>("keycloakMessage", textBody))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

        try {
            notifyClient.sendEmail(
                    "[notify_template_id]",
                    "[some_email_address]",
                    personalisation,
                    "[email_reference]");
        } catch (NotificationClientException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void close() {
    }
}

File: govuk-notify-spi/src/main/scala/uk/gov/nationalarchives/notifyspi/NotifyEmailSenderProvider.scala
package uk.gov.nationalarchives.notifyspi

import com.typesafe.config.{ConfigFactory, Config => TypeSafeConfig}
import org.keycloak.email.{EmailException, EmailSenderProvider}
import org.keycloak.models.UserModel
import software.amazon.awssdk.http.apache.ApacheHttpClient
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.ssm.SsmClient
import software.amazon.awssdk.services.ssm.model.GetParameterRequest
import uk.gov.service.notify.NotificationClient

import java.util
import scala.jdk.CollectionConverters._
import scala.util.{Failure, Success, Try}

class NotifyEmailSenderProvider() extends EmailSenderProvider {
  private val configFactory: TypeSafeConfig = ConfigFactory.load
  private val notifyApiKeyPath: String = configFactory.getString("notify.apiKeyPath")
  private val notifyTemplateIdPath: String = configFactory.getString("notify.templateIdPath")

  private def getApiKey: String = {
    getSsmParameterValue(notifyApiKeyPath)
  }

  private def getTemplateId: String = {
    getSsmParameterValue(notifyTemplateIdPath)
  }

  private def getSsmParameterValue(parameterPath: String): String = {
    val httpClient = ApacheHttpClient.builder.build
    val ssmClient: SsmClient = SsmClient.builder()
      .httpClient(httpClient)
      .region(Region.EU_WEST_2)
      .build()
    val getParameterRequest = GetParameterRequest.builder.name(parameterPath).withDecryption(true).build
    ssmClient.getParameter(getParameterRequest).parameter().value()
  }

  override def send(config: util.Map[String, String],
                     user: UserModel,
                     subject: String,
                     textBody: String,
                     htmlBody: String): Unit = {

    val notifyClient = new NotificationClient(getApiKey)

    val personalisation: Map[String, String] = Map(
      "keycloakMessage" -> textBody,
      "keycloakSubject" -> subject)

    sendNotifyEmail(notifyClient, NotifyEmailInfo(getTemplateId, user.getEmail, personalisation, user.getId))
  }

  override def close(): Unit = { }

  def sendNotifyEmail(notifyClient: NotificationClient, emailInfo: NotifyEmailInfo): Unit = {
    Try {
      notifyClient.sendEmail(
        emailInfo.templateId,
        emailInfo.userEmail,
        emailInfo.personalisation.asJava,
        emailInfo.reference)
    } match {
      case Failure(exception) => throw new EmailException(exception.getMessage)
      case Success(_) => ()
    }
  }

  override def send(config: util.Map[String, String], address: String, subject: String, textBody: String, htmlBody: String): Unit = ()
  override def validate(config: util.Map[String, String]): Unit = ()
}

case class NotifyEmailInfo(templateId: String, userEmail: String, personalisation: Map[String, String], reference: String)
